crates/typst-ide/src/matchers.rs

```diff
@@ -0,0 +1,178 @@
+                                    if let Some(res) = recv(NamedItem::Module(&value)) {
+                                        return Some(res);
+                                    }
+                                }
+                            } else if let Some(scope) = value.scope() {
+                                for (name, value) in scope.iter() {
+                                    let item = NamedItem::Import(
+                                        name,
+                                        Span::detached(),
+                                        Some(value),
+                                    );
+                                    if let Some(res) = recv(item) {
+                                        return Some(res);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                    Some(ast::Imports::Items(items)) => {
+                        for item in items.iter() {
+                            let name = item.bound_name();
+                            if let Some(res) =
+                                recv(NamedItem::Import(name.get(), name.span(), None))
+                            {
+                                return Some(res);
+                            }
+                        }
+                    }
+                }
+            }
+
+            sibling = node.prev_sibling();
+        }
+
+        if let Some(parent) = node.parent() {
+            if let Some(v) = parent.cast::<ast::ForLoop>() {
+                if node.prev_sibling_kind() != Some(SyntaxKind::In) {
+                    let pattern = v.pattern();
+                    for ident in pattern.bindings() {
+                        if let Some(res) = recv(NamedItem::Var(ident)) {
+                            return Some(res);
+                        }
+                    }
+                }
+            }
+
+            ancestor = Some(parent.clone());
+            continue;
+        }
+
+        break;
+    }
+
+    None
+}
+
+pub enum NamedItem<'a> {
+    Var(ast::Ident<'a>),
+    Fn(ast::Ident<'a>),
+    Module(&'a Module),
+    Import(&'a EcoString, Span, Option<&'a Value>),
+}
+
+impl<'a> NamedItem<'a> {
+    pub(crate) fn name(&self) -> &'a EcoString {
+        match self {
+            NamedItem::Var(ident) => ident.get(),
+            NamedItem::Fn(ident) => ident.get(),
+            NamedItem::Module(value) => value.name(),
+            NamedItem::Import(name, _, _) => name,
+        }
+    }
+
+    pub(crate) fn value(&self) -> Option<Value> {
+        match self {
+            NamedItem::Var(..) | NamedItem::Fn(..) => None,
+            NamedItem::Module(value) => Some(Value::Module((*value).clone())),
+            NamedItem::Import(_, _, value) => value.cloned(),
+        }
+    }
+}
+
+/// Find an expression that can be dereferenced on the given node.
```

Comment:
I'm not yet entirely clear on what exactly a "matcher" is and what the specific contract/idea of `deref_target` is.

Reply:
`deref_target` is a complete or incomplete "expression" node that can be operated by ide. It is a common matcher on AST that could be used by all ide functions, like `completion`, `tooltip` or `definition`. Two 



We can change the name of this function if we don't like it.

Reply:
`deref_target` is a complete or incomplete "expression" node that can be operated by ide. It is a common matcher on AST that could be used by all ide functions, like `completion`, `tooltip` or `definition`. Three advantages:
- make pattern matching in the ide functions more declarative and reusable:
  ```rs
  match deref_target(node) { 
    Callee(..) | VarAccess(..) => {}, // identifiers or identifier paths
    Label(..) => {}, // label expressions
    Normal(..) => {}, // normal expressions
    _ => ..,
  }
  ```
  v.s.
  ```rs
  if let ast::FuncCall(call) = node {
    // extract information from AST call
    do_identifier(..)
  } else if let ast::FieldAccess(fa) = node {
    do_identifier(..)
  } else if let ast::FieldAccess(fa) = node {
    do_label(..)
  } else if let ast::Error(err) = node {
    // recover errors
    do_identifier(..) or do_label(..)
  }
  ```
- resist changes of `typst_syntax` (AST).
- recover some common errors on imcomplete AST, like some opening labels.
  
We can change the name of this function if we don't like it.

Reply:
Well, the second version could also be written with match. But I see how it's nice to be able to process e.g. a FuncCall and a Set rule callee in a single match arm.